<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Core pipeline &mdash; 54gene-wgs-germline  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Release Notes" href="releasenotes.html" />
    <link rel="prev" title="Tests" href="tests.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> 54gene-wgs-germline
            <img src="_static/54gene_logo.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                1.0.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation and Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage and Execution</a></li>
<li class="toctree-l1"><a class="reference internal" href="investigateresults.html">Investigate the Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Tests</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Core pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#about">About</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#workflow">Workflow</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pulling-in-resources">Pulling in resources</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fastqc-and-read-trimming">FastQC and Read Trimming</a></li>
<li class="toctree-l4"><a class="reference internal" href="#read-alignment">Read Alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#variant-calling">Variant Calling</a></li>
<li class="toctree-l4"><a class="reference internal" href="#joint-genotyping">Joint genotyping</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="releasenotes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="fordevelopers.html">For developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">54gene-wgs-germline</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Core pipeline</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/corepipeline.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="core-pipeline">
<h1>Core pipeline<a class="headerlink" href="#core-pipeline" title="Permalink to this headline"></a></h1>
<section id="about">
<h2>About<a class="headerlink" href="#about" title="Permalink to this headline"></a></h2>
<p>This page describe details of the various run-modes available in this pipeline, the rules used within them and further specifications on the tools used. This page provides information on the parameters used for certain tools and behaviours between the run-modes.</p>
<section id="workflow">
<h3>Workflow<a class="headerlink" href="#workflow" title="Permalink to this headline"></a></h3>
<section id="pulling-in-resources">
<h4>Pulling in resources<a class="headerlink" href="#pulling-in-resources" title="Permalink to this headline"></a></h4>
<p>The pipeline first pulls in reference fasta, annotated VCF and index files (GRCh38 build) from the Broad Institute’s public AWS S3 bucket in the <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">get_resources</span></code>. The source code for this rule can be modified to pull from their GCP bucket instead.</p>
</section>
<section id="fastqc-and-read-trimming">
<h4>FastQC and Read Trimming<a class="headerlink" href="#fastqc-and-read-trimming" title="Permalink to this headline"></a></h4>
<p><strong>In ``full`` and ``fastqc_only`` run modes::</strong>
The pipeline will create symbolic links for the supplied input fastqs from the <code class="docutils literal notranslate"><span class="pre">manifest.txt</span></code> file in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">symlink_fastqs</span></code> using a module from <code class="docutils literal notranslate"><span class="pre">utilities.py</span></code> to obtain the readgroup portion of the sample names for ease of naming files thereon.</p>
<p><code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">fastqc</span></code> will then generate fastQC reports for all input fastqs followed by read trimming and adapter removal (Illumina’s TruSeq adapters) using <code class="docutils literal notranslate"><span class="pre">fastp</span></code>.</p>
<p>Post-trimming, fastQC will be run again on the trimmed reads. This results in fastQC results for the raw input reads in <code class="docutils literal notranslate"><span class="pre">results/fastqc</span></code>, and post-trimmed reads in <code class="docutils literal notranslate"><span class="pre">results/post_trimming_fastqc</span></code>.</p>
</section>
<section id="read-alignment">
<h4>Read Alignment<a class="headerlink" href="#read-alignment" title="Permalink to this headline"></a></h4>
<p>Trimmed reads are aligned to the reference genome using BWA in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">align_reads</span></code>. The read1 and read2 data are combined into a single output BAM per fastq pair. If samples were run over several lanes (i.e. 4 lanes), the read1 and read2 fastq pair will be aligned separately then combined during a deduplication step later on (see <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">mark_duplicates</span></code>). This helps speed up the alignment process and helps parallelize this step for efficiency.</p>
<p>Duplicates for each sample BAM are the marked and removed. Multi-lane sample BAMs will be concatenated in this step. The pipeline will then use GATK’s BaseRecalibrator to generate a recalibration table for base quality score using known sites VCF pulled from the Broad’s resources bucket. This recalibration is then applied to each BAM in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">apply_bqsr</span></code>. Samtools stats are then generated for all BAMs.</p>
</section>
<section id="variant-calling">
<h4>Variant Calling<a class="headerlink" href="#variant-calling" title="Permalink to this headline"></a></h4>
<p>Single-sample gVCFs are generated using GATK’s HaplotypeCaller and parallelized over a user-specified number of intervals and/or regions of the genome using the <code class="docutils literal notranslate"><span class="pre">intervals.tsv</span></code> file. A gVCF for each sample will be generated for each specified interval/region. This method allows for parallelization and reduction in overall execution time for variant calling. Following the GVCF workflow, these are to be used for joint genotyping of multiple samples later in the pipeline for scalable analyses.</p>
<p>Each gVCF will be compressed, indexed and then concatenated across each interval/region to generate a single gVCF for each sample with the naming convention of <code class="docutils literal notranslate"><span class="pre">/HaplotypeCaller/called/{sample}_all_regions.g.vcf.gz</span></code>.</p>
</section>
<section id="joint-genotyping">
<h4>Joint genotyping<a class="headerlink" href="#joint-genotyping" title="Permalink to this headline"></a></h4>
<p><strong>In \``joint_genotyping``\ mode only:</strong>
It is at this step in the workflow that a secondary entry point is provided when the run-mode in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> is set to <code class="docutils literal notranslate"><span class="pre">joint_genotyping</span></code>. In this run mode, the gVCFs provided in the <code class="docutils literal notranslate"><span class="pre">manifest.txt</span></code> file and their indexes will be symlinked prior to continuing on to the rest of the workflow.</p>
<p><strong>In ``joint_genotyping`` and ``full`` run modes:</strong>
In order to perform joint-genotyping over multiple samples using GATK’s GenotypeGVCFs, the input gVCFs must be consolidated across samples as the genotyping step can only take one single input. To circumvent this issue, we use GATK’s GenomicsDBImport in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">HC_consolidate_gvcfs</span></code> to generate database stores for each sample, parallelized again across intervals/regions, to then pass into GenotypeGVCFs. DBImport can potentially take up alot of /temp space so it is recommended that <code class="docutils literal notranslate"><span class="pre">--tmp-dir</span></code> be used to redirect to a larger temp space. The <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code> and <code class="docutils literal notranslate"><span class="pre">--reader-threads</span></code> parameters can be tweaked in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> to read in more data stores concurrently or in larger batch sizes but the default settings are those suggested by GATK developers.</p>
<p>Joint genotyping using the various database stores created is performed in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">HC_genotype_gvcfs</span></code> to emit a genotyped gVCF for each interval/region in <code class="docutils literal notranslate"><span class="pre">results/HaplotypeCaller/genotyped/{interval}.vcf.gz</span></code>. The <code class="docutils literal notranslate"><span class="pre">--max_alt_alleles</span></code> to genotype and <code class="docutils literal notranslate"><span class="pre">--max_genotype_count</span></code> for each site can be tweaked in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
<p>The subsequent per-interval/region, genotyped gVCFs are concatenated into one project-level multi-sample gVCF for downstream analysis in <code class="docutils literal notranslate"><span class="pre">results/HaplotypeCaller/genotyped/HC_variants.vcf.gz</span></code>.</p>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tests.html" class="btn btn-neutral float-left" title="Tests" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="releasenotes.html" class="btn btn-neutral float-right" title="Release Notes" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, B. Ballew, C. Palmer, E. Joshi, 54gene.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>