<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Core pipeline &mdash; 54gene-wgs-germline  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Release Notes" href="releasenotes.html" />
    <link rel="prev" title="Tests" href="tests.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

<a href="index.html" class="icon icon-home"> 54gene-wgs-germline

<a href="https://54gene.com">
    <img src="_static/54gene_logo.png" class="logo" alt="Logo"/>
  </a>


  
  
  
    <div class="version">
      1.0.0
    </div>
  



        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation and Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage and Execution</a></li>
<li class="toctree-l1"><a class="reference internal" href="investigateresults.html">Investigate the Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Tests</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Core pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#about">About</a></li>
<li class="toctree-l2"><a class="reference internal" href="#workflow">Workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pulling-in-resources">Pulling in resources</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastqc-and-read-trimming">FastQC and Read Trimming</a></li>
<li class="toctree-l3"><a class="reference internal" href="#read-alignment">Read Alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="#variant-calling">Variant Calling</a></li>
<li class="toctree-l3"><a class="reference internal" href="#joint-genotyping">Joint genotyping</a></li>
<li class="toctree-l3"><a class="reference internal" href="#variant-filtration">Variant Filtration</a></li>
<li class="toctree-l3"><a class="reference internal" href="#post-calling-qc">Post-Calling QC</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#contamination-check">Contamination Check</a></li>
<li class="toctree-l4"><a class="reference internal" href="#merging-calls">Merging calls</a></li>
<li class="toctree-l4"><a class="reference internal" href="#checking-relatedness-with-somalier">Checking Relatedness with Somalier</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sex-check">Sex Check</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sample-exclusions">Sample Exclusions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multiqc">MultiQC</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="releasenotes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="fordevelopers.html">For developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">54gene-wgs-germline</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Core pipeline</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/corepipeline.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="core-pipeline">
<h1>Core pipeline<a class="headerlink" href="#core-pipeline" title="Permalink to this heading"></a></h1>
<section id="about">
<h2>About<a class="headerlink" href="#about" title="Permalink to this heading"></a></h2>
<p>This page describe details of the various run-modes available in this pipeline, the rules used within them and further specifications on the tools used. This page provides information on the parameters used for certain tools and behaviours between the run-modes.</p>
</section>
<section id="workflow">
<h2>Workflow<a class="headerlink" href="#workflow" title="Permalink to this heading"></a></h2>
<section id="pulling-in-resources">
<h3>Pulling in resources<a class="headerlink" href="#pulling-in-resources" title="Permalink to this heading"></a></h3>
<p>The pipeline first pulls in reference fasta, annotated VCF and index files (GRCh38 build) from the Broad Institute’s public AWS S3 bucket in the <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">get_resources</span></code>. The source code for this rule can be modified to pull from their GCP bucket instead.</p>
</section>
<section id="fastqc-and-read-trimming">
<h3>FastQC and Read Trimming<a class="headerlink" href="#fastqc-and-read-trimming" title="Permalink to this heading"></a></h3>
<p><strong>In full and fastqc_only run modes:</strong></p>
<p>The pipeline will create symbolic links for the supplied input fastqs from the <code class="docutils literal notranslate"><span class="pre">manifest.txt</span></code> file in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">symlink_fastqs</span></code> using a module from <code class="docutils literal notranslate"><span class="pre">utilities.py</span></code> to obtain the readgroup portion of the sample names for ease of file naming of FastQC outputs which cannot be renamed and of outputs later on as well. Please bare in mind these symlinks when managing data on your infrastructure.</p>
<p><code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">fastqc</span></code> will then generate fastQC reports for all input fastqs followed by read trimming and adapter removal (Illumina’s TruSeq adapters) using <code class="docutils literal notranslate"><span class="pre">fastp</span></code>. If you choose to use alternate adapters, you will need to modify the source code for this particular rule. There also exist many additional parameters and functionality for fastp which we have not exposed in the config or included here.</p>
<p>Why fastp over other standard tools such as Trimmomatic?
We chose fastp over Trimmomatic after encountering run errors and periodic stalling of the tool in our initial development of the pipeline.</p>
<p>Post-trimming, fastQC will be run again on the trimmed reads. This results in fastQC results for the raw input reads in <code class="docutils literal notranslate"><span class="pre">results/fastqc</span></code>, and post-trimmed reads in <code class="docutils literal notranslate"><span class="pre">results/post_trimming_fastqc</span></code>.</p>
</section>
<section id="read-alignment">
<h3>Read Alignment<a class="headerlink" href="#read-alignment" title="Permalink to this heading"></a></h3>
<p><strong>In full mode only:</strong></p>
<p>Trimmed reads are aligned to the reference genome using BWA in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">align_reads</span></code>. The read1 and read2 data are combined into a single output BAM per fastq pair. If samples were run over several lanes (i.e. 4 lanes), the read1 and read2 fastq pair will be aligned separately then combined during a deduplication step later on (see <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">mark_duplicates</span></code>). This helps speed up the alignment process and helps parallelize this step for efficiency. The output of this rule will be a sorted BAM for each sample within <code class="docutils literal notranslate"><span class="pre">results/mapped/{rg}.bam</span></code>.</p>
<p>Duplicates for each sorted sample BAM are the marked and removed. Multi-lane sample BAMs will be concatenated in this step. The pipeline will then use GATK’s BaseRecalibrator to generate a recalibration table for base quality score using known sites VCF pulled from the Broad’s resources bucket. This recalibration is then applied to each BAM in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">apply_bqsr</span></code>. Samtools stats are then generated for all BAMs. See the <a class="reference internal" href="investigateresults.html"><span class="doc">Investigate the Results</span></a> page for more information on where to find the stats.</p>
<p>We are currently using the dbSNP VCFs of known SNPs and the known indels VCF from the GATK resource bundle, available on <a class="reference external" href="s3://broad-references/hg38/v0/">AWS S3</a> and <a class="reference external" href="https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0">GCP</a>.</p>
</section>
<section id="variant-calling">
<h3>Variant Calling<a class="headerlink" href="#variant-calling" title="Permalink to this heading"></a></h3>
<p>Single-sample gVCFs are generated using GATK’s HaplotypeCaller and parallelized over a user-specified number of intervals and/or regions of the genome using the <code class="docutils literal notranslate"><span class="pre">intervals.tsv</span></code> file. A gVCF for each sample will be generated for each specified interval/region. This method allows for parallelization and reduction in overall execution time for variant calling. Following the <a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller">GVCF workflow</a>, these are to be used for joint genotyping of multiple samples later in the pipeline for scalable analyses.</p>
<p>Each gVCF will be compressed and then concatenated across each interval/region to generate a single gVCF for each sample along with its index, with the naming convention of <code class="docutils literal notranslate"><span class="pre">/HaplotypeCaller/called/{sample}_all_regions.g.vcf.gz</span></code>.</p>
</section>
<section id="joint-genotyping">
<h3>Joint genotyping<a class="headerlink" href="#joint-genotyping" title="Permalink to this heading"></a></h3>
<p><strong>In joint_genotyping mode only:</strong></p>
<p>It is at this step in the workflow that a secondary entry point is provided when the run-mode in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> is set to <code class="docutils literal notranslate"><span class="pre">joint_genotyping</span></code>. In this run mode, the gVCFs provided in the <code class="docutils literal notranslate"><span class="pre">manifest.txt</span></code> file and their indexes will be symlinked to a subdirectory within <code class="docutils literal notranslate"><span class="pre">/results/HaplotypeCaller/called/</span></code> prior to continuing on to the rest of the workflow.</p>
<p><strong>In joint_genotyping and full run modes:</strong></p>
<p>In order to perform joint-genotyping over multiple samples using GATK’s GenotypeGVCFs, the input gVCFs must be consolidated across samples as the genotyping step can only take one single input. To circumvent this issue, we use GATK’s GenomicsDBImport in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">HC_consolidate_gvcfs</span></code> to generate database stores for each sample, parallelized again across intervals/regions, to then pass into GenotypeGVCFs. DBImport can potentially take up alot of /temp space so it is recommended that <code class="docutils literal notranslate"><span class="pre">--tmp-dir</span></code> be used to redirect to a larger temp space. The <code class="docutils literal notranslate"><span class="pre">--batch-size</span></code> and <code class="docutils literal notranslate"><span class="pre">--reader-threads</span></code> parameters can be tweaked in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> to read in more data stores concurrently or in larger batch sizes but the default settings are those suggested by GATK developers.</p>
<p>Joint genotyping using the various database stores created is performed in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">HC_genotype_gvcfs</span></code> to emit a genotyped gVCF for each interval/region in <code class="docutils literal notranslate"><span class="pre">results/HaplotypeCaller/genotyped/{interval}.vcf.gz</span></code>. The <code class="docutils literal notranslate"><span class="pre">--max_alt_alleles</span></code> to genotype and <code class="docutils literal notranslate"><span class="pre">--max_genotype_count</span></code> for each site can be tweaked in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
<p>We exposed these and other parameters for GenomicsDBImport after encountering <a class="reference external" href="https://github.com/broadinstitute/gatk/issues/7542">recent issues</a> where the <code class="docutils literal notranslate"><span class="pre">--max-alternate-alleles</span></code> flag for GenotypeGVCFs was set at a default of 6 but was not actually being applied as a threshold. A fix in GATK v4.2.4.1 attempted to apply this threshold instead resulted in a bug where the relevant locus would result in the program crashing. Subsequently, an <a class="reference external" href="https://github.com/broadinstitute/gatk/pull/7655">update in v4.2.5.0</a> introduced a new parameter for GenotypeGVCFs called <code class="docutils literal notranslate"><span class="pre">--genomicsdb-max-alternate-alleles</span></code> and required to be minimum one greater than <code class="docutils literal notranslate"><span class="pre">--max-alternate-alleles</span></code> to account for the NON_REF allele.</p>
<p>The per-interval/region, genotyped gVCFs will be concatenated into one sorted, indexed, project-level multi-sample gVCF for downstream analysis in <code class="docutils literal notranslate"><span class="pre">results/HaplotypeCaller/genotyped/HC_variants.vcf.gz</span></code>.</p>
<p><em>Note</em>: While GenomicsDBImport supports adding N+1 samples to the datastores, however, our pipeline does not utilize this functionality and instead creates the databases every time from scratch. This was a development choice made to avoid issues with potential failures with maintaining the datastores and revisitng them in future analyses.</p>
</section>
<section id="variant-filtration">
<span id="id1"></span><h3>Variant Filtration<a class="headerlink" href="#variant-filtration" title="Permalink to this heading"></a></h3>
<p>The project-level VCF is normalized and split into a separate VCF with multiallelics to allow for further hard-filtering into SNPs and indels later. Hard-filtering using GATK’s VariantFiltration tool is performed separately on the SNP and indel-specific project-level VCFs in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">hard_filter_snps</span></code> and <code class="docutils literal notranslate"><span class="pre">rule_hard_filter_indels</span></code>. For more information on how we perform hard-filtering, see GATK’s documentation on hard-filtering recommendations <a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890471-Hard-filtering-germline-short-variants">here</a>.</p>
<p><em>Note</em>: We currently do not remove the filtered sites themselves from the VCF but instead just update the filter field.</p>
</section>
<section id="post-calling-qc">
<h3>Post-Calling QC<a class="headerlink" href="#post-calling-qc" title="Permalink to this heading"></a></h3>
<section id="contamination-check">
<h4>Contamination Check<a class="headerlink" href="#contamination-check" title="Permalink to this heading"></a></h4>
<p><strong>In full mode only:</strong></p>
<p>As an added QC measure, we perform a contamination check on the BAM files using a tool called <a class="reference external" href="https://genome.sph.umich.edu/wiki/VerifyBamID">VerifyBamID</a>. This tool will verify whether reads in a particular BAM are concordant with previous genotypes found for a specific sample. This can help identify potential sample mixture.</p>
<p>The tool normally takes the entire BAM file as an input but to reduce the computational burden of performing this check, we opted to only subset particular chromosomes (ideally one or two) from the BAM files to perform the check. We have found that is this sufficient for initial flagging of contamination for further ni-depth investigation of troublesome samples. We allow the ability to select these chromosomes within the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
<p>This step in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">contamination_check</span></code> will output various contamination metrics for each sample BAM file that are combined in a summary file. This summary file will be later used for automated filtering of samples out of the project-level VCF based on thresholds defined in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>. See <a class="reference internal" href="#exclude-samples"><span class="std std-ref">Sample Exclusions</span></a> section for more information.</p>
</section>
<section id="merging-calls">
<h4>Merging calls<a class="headerlink" href="#merging-calls" title="Permalink to this heading"></a></h4>
<p>The hard-filtered SNP and indels generated in <a class="reference internal" href="#variant-filtration"><span class="std std-ref">Variant Filtration</span></a> are merged into one sorted, and indexed project-level hard-filtered multi-sample VCF in <code class="docutils literal notranslate"><span class="pre">results/HaplotypeCaller/filtered/HC_variants.hardfiltered.vcf.gz</span></code>. We subsequnently run GATK’s CollectVariantCallMetrics on this VCF file. See <a class="reference internal" href="investigateresults.html"><span class="doc">Investigate the Results</span></a> for more info.</p>
</section>
<section id="checking-relatedness-with-somalier">
<h4>Checking Relatedness with Somalier<a class="headerlink" href="#checking-relatedness-with-somalier" title="Permalink to this heading"></a></h4>
<p>If specified as ‘Yes’ for <code class="docutils literal notranslate"><span class="pre">check_relatedness</span></code> in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>, the pipeline will run Somalier to check for relatedness amongst the samples. <a class="reference external" href="https://github.com/brentp/somalier">Somalier</a> is a tool that can be used to check any number of samples from joint-called VCFs for identity and to infer relationships. The tool ideally requires a jointly-called cohort VCF and PED file of expected sexes and relationships. An example of the Somalier output can be found <a class="reference external" href="https://brentp.github.io/somalier/ex.html">here</a>.</p>
<p>In our pipeline, we create a PED-style file from the <code class="docutils literal notranslate"><span class="pre">sex_linker.tsv</span></code> specified in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> using a module that will convert the linker file to the appropriate format for Somalier in <code class="docutils literal notranslate"><span class="pre">utilities/create_ped.py</span></code>.</p>
<p>How do we use this tool?</p>
<p>We use this tool as a rough estimate of relatedness and to highlight potential concerns within our data for putative relatedness or with unexpected duplicates. If identified, we then perform more thorough analyses using other tools such as KING, plink, graf and etc.</p>
<p>This is largely because Somalier uses the following equation to determine relatedness:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="n">shared</span><span class="o">-</span><span class="n">hets</span><span class="p">)(</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">)</span><span class="o">-</span><span class="mi">2</span><span class="o">*</span><span class="n">ibs0</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">)</span><span class="o">/</span><span class="nb">min</span> <span class="p">(</span><span class="n">hets</span><span class="p">(</span><span class="n">i</span><span class="p">),</span><span class="n">hets</span><span class="p">(</span><span class="n">j</span><span class="p">))</span>
</pre></div>
</div>
<p>This assumes, as noted in their <a class="reference external" href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-020-00761-2">publication</a>, that the sites they’ve selected on which to assess relatedness are “high-quality, unlinked sites with a population allele frequency of around 0.5.”  We suspect this will not hold true across all populations. It is unclear how much this will degrade across multiple populations with some degree of shared ancestry.  Note that the relatedness value will always be depressed when comparing, for example, NA12878 with Nigerian subjects.</p>
<p><em>Note</em>: We currently run this tool using our organization’s conda channel. You will likely have to modify the <code class="docutils literal notranslate"><span class="pre">envs/somalier.yaml</span></code> to point elsewhere to a different package location.</p>
</section>
<section id="sex-check">
<h4>Sex Check<a class="headerlink" href="#sex-check" title="Permalink to this heading"></a></h4>
<p>In addition to Somalier above, we also use bcftools’ guess-ploidy to determine sample sex from genotype likelihoods. These results are included in the MultiQC report generated at the end of the post-calling QC stage. See <a class="reference internal" href="#multiqc"><span class="std std-ref">MultiQC</span></a> for more information.</p>
</section>
<section id="sample-exclusions">
<span id="exclude-samples"></span><h4>Sample Exclusions<a class="headerlink" href="#sample-exclusions" title="Permalink to this heading"></a></h4>
<p>We exclude samples from the project-level hard-filtered VCF using a module found in <code class="docutils literal notranslate"><span class="pre">script/create_exclude_list.py</span></code> in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">create_exclude_list</span></code> that will use the metrics and information generated from the contamination check and bcftools stats to exclude samples based on the following default thresholds:</p>
<ul class="simple">
<li><p>Max het/hom ratio of 2.5</p></li>
<li><p>Minimum average depth of 20</p></li>
<li><p>Maximum contamination estimate of 0.03 (only used if run in full run mode)</p></li>
</ul>
<p>These thresholds can be tweaked in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>. A list of samples to exclude and another list with these samples and annotations for why they were excluded will be generated in <code class="docutils literal notranslate"><span class="pre">results/post_qc_exclusions/</span></code>.</p>
<p>Post sample exclusion, another sorted and indexed, project-level, hard-filtered VCF will emitted in <code class="docutils literal notranslate"><span class="pre">results/post_qc_exclusions/samples_excluded.HC_variants.hardfiltered.vcf.gz</span></code>. This is to avoid altering the original project-level, hard-filtered VCF. Note that the ID column here will also be updated to <code class="docutils literal notranslate"><span class="pre">%CHROM:%POS:%REF:%ALT</span></code> using bcftools annotate. If you want this field to updated to something else, you may choose to modify the source code in <code class="docutils literal notranslate"><span class="pre">rule</span> <span class="pre">exclude_samples</span></code>.</p>
</section>
<section id="multiqc">
<span id="id3"></span><h4>MultiQC<a class="headerlink" href="#multiqc" title="Permalink to this heading"></a></h4>
<p>A multiQC report is generated for all three run-modes and will differ in content depending on which post-calling QC checks were performed.</p>
<p>For <strong>fastqc_only</strong> run mode, the multiQC report will include:</p>
<ul class="simple">
<li><p>Pre- and post-read-trimming fastQC results</p></li>
</ul>
<p>For the <strong>full</strong> run mode, the multiQC report will include:</p>
<ul class="simple">
<li><p>Pre- and post-read-trimming fastQC results</p></li>
<li><p>Bcftool stats on joint-called variants</p></li>
<li><p>Deduplication metrics for BAM files</p></li>
<li><p>Sex check results from bcftools guess-ploidy</p></li>
<li><p>Contamination check results from verifyBamID</p></li>
<li><p>If specified in config, relatedness check results from Somalier</p></li>
<li><p>Variant calling metrics</p></li>
</ul>
<p>For <strong>joint_genotyping</strong> mode, the multiQC report will include:</p>
<ul class="simple">
<li><p>Variant calling metrics</p></li>
<li><p>Sex check results from bcftools guess-ploidy</p></li>
<li><p>Bcftool stats on joint-called variants</p></li>
<li><p>If specified in config, relatedness check results from Somalier</p></li>
</ul>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
          <a href="tests.html" class="btn btn-neutral float-left" title="Tests" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
          <a href="releasenotes.html" class="btn btn-neutral float-right" title="Release Notes" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
      </div>
  
    <hr/>
  
    <div role="contentinfo">
      <p>&#169; Copyright 2022, B. Ballew, C. Palmer, E. Joshi, 54gene.</p>
    </div>
  
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a>
      
     
  </footer>
  
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>