<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Overview &mdash; 54gene-wgs-germline  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->

        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="installation.html" />
    <link rel="prev" title="54gene Whole Genome Sequencing Germline Calling Pipeline" href="index.html" />
</head>

<body class="wy-body-for-nav">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

<a href="index.html" class="icon icon-home"> 54gene-wgs-germline
<a href="https://54gene.com">
    <img src="_static/54gene_logo.png" class="logo" alt="Logo"/>
</a>
  <div class="version">
    1.0.0
  </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="execution.html">Execution</a></li>
<li class="toctree-l1"><a class="reference internal" href="investigateresults.html">Investigate the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Tests</a></li>
<li class="toctree-l1"><a class="reference internal" href="corepipeline.html">Core pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="fordevelopers.html">For developers</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">54gene-wgs-germline</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Overview</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/overview.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">

  <section id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this heading"></a></h1>
<p>This workflow was designed by the Genomics &amp; Data Science team (GDS) at 54gene and is used to analyze paired-end short-read germline whole-genome sequencing data.  This pipeline is designed to first be deployed in small batches (e.g. per flow cell), starting with FASTQs and resulting in gVCFs and a small batch joint-called VCF.  A second run of the pipeline can receive a larger batch of gVCFs (e.g. gVCFs derived from many flow cells), and generates a large batch joint-called VCF.  The workflow, which is designed to support reproducible bioinformatics, is written in <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/">Snakemake</a> and is platform-agnostic.  All dependencies are installed by the pipeline as-needed using <a class="reference external" href="https://docs.conda.io/en/latest/">conda</a>.  Development and testing has been predominantly on AWS’ <a class="reference external" href="https://aws.amazon.com/hpc/parallelcluster/">ParallelCluster</a> using <a class="reference external" href="https://aws.amazon.com/amazon-linux-2/">Amazon Linux</a> using Snakemake version 7.8.2.</p>
<p>Features:</p>
<ul class="simple">
<li><p>Read filtering and trimming</p></li>
<li><p>Read alignment, deduplication, and BQSR</p></li>
<li><p>Variant calling and filtering</p></li>
<li><p>Joint-genotyping</p></li>
<li><p>Sex discordance and relatedness assessment</p></li>
<li><p>Generate MultiQC reports</p></li>
</ul>
<p>To install the latest release, type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">clone</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">gitlab</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">data</span><span class="o">-</span><span class="n">analysis5</span><span class="o">/</span><span class="n">dna</span><span class="o">-</span><span class="n">sequencing</span><span class="o">/</span><span class="mi">54</span><span class="n">gene</span><span class="o">-</span><span class="n">wgs</span><span class="o">-</span><span class="n">germline</span><span class="o">.</span><span class="n">git</span>
</pre></div>
</div>
<section id="inputs">
<h2>Inputs<a class="headerlink" href="#inputs" title="Permalink to this heading"></a></h2>
<p>The pipeline requires the following inputs:</p>
<ul class="simple">
<li><p>A headerless, whitespace delimited <code class="docutils literal notranslate"><span class="pre">manifest.txt</span></code> file with sample names and paths (columns dependent on the run-mode)</p></li>
<li><p>Config file with the run-mode specified and other pipeline parameters configured (see default config provided in <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code>)</p></li>
<li><p>A tab-delimited <code class="docutils literal notranslate"><span class="pre">intervals.tsv</span></code> file with names of intervals and paths to region (BED) files of the genome you want to parallelize the variant calling and joint-calling steps by (i.e. 50 BED files each with a small region of the genome to parallelize by)</p></li>
<li><p>A tab-delimited <code class="docutils literal notranslate"><span class="pre">sex_linker.tsv</span></code> file with the sample names in one column and sex in the other to identify discordances in reported vs. inferred sex</p></li>
<li><p>A <code class="docutils literal notranslate"><span class="pre">multiqc.yaml</span></code> config file for generating MultiQC reports (provided for you)</p></li>
</ul>
</section>
<section id="outputs">
<h2>Outputs<a class="headerlink" href="#outputs" title="Permalink to this heading"></a></h2>
<p>Depending on which run-mode you have set, you will be able to generate:</p>
<ul class="simple">
<li><p>A hard-filtered, multi-sample joint-called VCF in <code class="docutils literal notranslate"><span class="pre">full</span></code> and <code class="docutils literal notranslate"><span class="pre">joint_genotyping</span></code> mode</p></li>
<li><p>Per-sample gVCFs for all regions of the genome for future joint-calling in <code class="docutils literal notranslate"><span class="pre">full</span></code> mode</p></li>
<li><p>Deduplicated and post-BQSR BAM files in <code class="docutils literal notranslate"><span class="pre">full</span></code> mode</p></li>
<li><p>Various QC metrics (e.g. FastQC, MultiQC, bcftools stats) in all three modes</p></li>
</ul>
<p>See the <a class="reference internal" href="installation.html"><span class="doc">Installation</span></a>, <a class="reference internal" href="execution.html"><span class="doc">Execution</span></a>, and <a class="reference internal" href="configuration.html"><span class="doc">Configuration</span></a> for details on setting up and running the pipeline.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
          <a href="index.html" class="btn btn-neutral float-left" title="54gene Whole Genome Sequencing Germline Calling Pipeline" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
          <a href="installation.html" class="btn btn-neutral float-right" title="Installation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
      </div>

    <hr/>

    <div role="contentinfo">
      <p>&#169; Copyright 2022, B. Ballew, E. Joshi, C. Palmer, 54gene.</p>
    </div>

    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a>


  </footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

</body>
</html>
