<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage and Execution &mdash; 54gene-wgs-germline  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Investigate the Results" href="investigateresults.html" />
    <link rel="prev" title="Installation and Setup" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> 54gene-wgs-germline
            <img src="_static/54gene_logo.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                1.0.0
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation and Setup</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage and Execution</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#config-parameters">Config Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#resource-allocation">Resource Allocation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#execution">Execution</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#deploying-the-pipeline">Deploying the pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#logging">Logging</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="investigateresults.html">Investigate the Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Tests</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="corepipeline.html">Core pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="releasenotes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="fordevelopers.html">For developers</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">54gene-wgs-germline</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Usage and Execution</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/usage.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="usage-and-execution">
<h1>Usage and Execution<a class="headerlink" href="#usage-and-execution" title="Permalink to this headline"></a></h1>
<section id="config-parameters">
<h2>Config Parameters<a class="headerlink" href="#config-parameters" title="Permalink to this headline"></a></h2>
<p>Below are descriptions and usage options for the various config parameters specified in <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 22%" />
<col style="width: 16%" />
<col style="width: 23%" />
<col style="width: 39%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Required</p></th>
<th class="head"><p>Default value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sampleFile</p></td>
<td><p>Y</p></td>
<td><p>NA</p></td>
<td><p>Manifest file with IDs</p></td>
</tr>
<tr class="row-odd"><td><p>intervalsFile</p></td>
<td><p>Y</p></td>
<td><p>NA</p></td>
<td><p>File with interval names
and file paths</p></td>
</tr>
<tr class="row-even"><td><p>jobs</p></td>
<td><p>Y</p></td>
<td><p>1000</p></td>
<td><p>Jobs to submit to cluster</p></td>
</tr>
<tr class="row-odd"><td><p>sexLinker</p></td>
<td><p>Y</p></td>
<td><p>NA</p></td>
<td><p>File with reported sex of
sexes of each sample ID</p></td>
</tr>
<tr class="row-even"><td><p>tempDir</p></td>
<td><p>Y</p></td>
<td><p>temp/</p></td>
<td><p>Location of temp directory</p></td>
</tr>
<tr class="row-odd"><td><p>runType</p></td>
<td><p>Y</p></td>
<td><p>full: yes</p></td>
<td><p>Specify run mode to use</p></td>
</tr>
<tr class="row-even"><td><p>global_vars</p></td>
<td><p>N</p></td>
<td><p>As specified</p></td>
<td><p>Set global java options</p></td>
</tr>
<tr class="row-odd"><td><p>cluster_mode</p></td>
<td><p>N</p></td>
<td><p>As specified</p></td>
<td><p>Used to submit jobs to a
cluster. See
<a class="reference internal" href="#execution"><span class="std std-ref">Execution</span></a></p></td>
</tr>
<tr class="row-even"><td><p>default_queue</p></td>
<td><p>Y</p></td>
<td><p>“big”</p></td>
<td><p>Name of your default
cluster partition/queue</p></td>
</tr>
<tr class="row-odd"><td><p>compute_queue</p></td>
<td><p>Y</p></td>
<td><p>“big”</p></td>
<td><p>Name of queue/partition
with compute-heavy nodes</p></td>
</tr>
<tr class="row-even"><td><p>memory_queue</p></td>
<td><p>Y</p></td>
<td><p>“big”</p></td>
<td><p>Name of queue/partition
with nodes for memory-
intensive jobs</p></td>
</tr>
<tr class="row-odd"><td><p>bed</p></td>
<td><p>N</p></td>
<td><p><cite>Homo_sapiens_
assembly38.bed</cite></p></td>
<td><p>Split genome by chromosome
when calling variants</p></td>
</tr>
<tr class="row-even"><td><p>max_concurent</p></td>
<td><p>Y</p></td>
<td><p>40</p></td>
<td><p>Max concurrent running jobs
to run at once. Allows you
to throttle heavy rules
depending on your env</p></td>
</tr>
<tr class="row-odd"><td><p>max_het_ratio</p></td>
<td><p>Y</p></td>
<td><p>2.5</p></td>
<td><p>Max het/hom ratio to allow</p></td>
</tr>
<tr class="row-even"><td><p>min_avg_depth</p></td>
<td><p>Y</p></td>
<td><p>20</p></td>
<td><p>Minimum depth required for
sample</p></td>
</tr>
<tr class="row-odd"><td><p>max_contam</p></td>
<td><p>Y</p></td>
<td><p>0.03</p></td>
<td><p>Max % of contam allowed</p></td>
</tr>
<tr class="row-even"><td><p>time_threshold</p></td>
<td><p>Y</p></td>
<td><p>5 (mins)</p></td>
<td><p>Exclude rules from the
benchmarking report if
elapsed time is below this</p></td>
</tr>
<tr class="row-odd"><td><p>somalier</p></td>
<td><p>Y</p></td>
<td><p>True</p></td>
<td><p>Check relatedness and sex
discordance with Somalier
(requires sex_linker.tsv)
only available in full
run mode</p></td>
</tr>
</tbody>
</table>
</section>
<section id="resource-allocation">
<h2>Resource Allocation<a class="headerlink" href="#resource-allocation" title="Permalink to this headline"></a></h2>
<p>This pipeline was originally developed to be run on an Unix-based HPC system for scalable and efficient analyses. As a result, there are several configuration options available for allocating resources and running jobs within the workflow.</p>
<p>Within the <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code> there are several options for allocating memory, threads, and setting Java-specific variables (for GATK suite of tools) for the various tools used in the workflow.</p>
<ul class="simple">
<li><p>To modify the number of maximum jobs to run, you can specify a number for <code class="docutils literal notranslate"><span class="pre">jobs</span></code></p></li>
<li><p>Set a maximum number of jobs to run concurrently if you have bandwidth constraints using the <code class="docutils literal notranslate"><span class="pre">max_concurrent</span></code> variable</p></li>
<li><p>Specify which paritions/queues to submit your jobs to, depending on your HPC using the <code class="docutils literal notranslate"><span class="pre">default_queue</span></code>, <code class="docutils literal notranslate"><span class="pre">compute_queue</span></code> and <code class="docutils literal notranslate"><span class="pre">memory_queue</span></code> variables (some jobs such as alignment, require more memory and can use nodes with more memory available when specified with these variables)</p></li>
<li><p>Specify the number of threads and memory in MB for each tool, where available using the <code class="docutils literal notranslate"><span class="pre">threads</span></code> and <code class="docutils literal notranslate"><span class="pre">memory</span></code> variables</p></li>
<li><p>Specify the space to allocate for Java class metadata using the <code class="docutils literal notranslate"><span class="pre">global_vars</span></code> variable</p></li>
</ul>
<section id="execution">
<span id="id1"></span><h3>Execution<a class="headerlink" href="#execution" title="Permalink to this headline"></a></h3>
</section>
</section>
<section id="deploying-the-pipeline">
<h2>Deploying the pipeline<a class="headerlink" href="#deploying-the-pipeline" title="Permalink to this headline"></a></h2>
<p>With the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> configured to your run-mode of choice with paths to the necessary manifest and input files, the workflow can be executed on any infrastructure using the <code class="docutils literal notranslate"><span class="pre">snakemake</span></code> command, supplied with further Snakemake command-line arguments (i.e. specifying a profile with <code class="docutils literal notranslate"><span class="pre">--profile</span></code> or <code class="docutils literal notranslate"><span class="pre">--cluster</span></code> to submit jobs to an HPC) depending on your environment.</p>
<p>Fr example, you may execute the workflow on a cluster using something like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snakemake</span> <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">conda</span> <span class="o">--</span><span class="n">cluster</span> <span class="n">qsub</span> <span class="o">--</span><span class="n">jobs</span> <span class="mi">100</span>
</pre></div>
</div>
<p>The pipeline is configured to automatically create a subdirectory for logs in <code class="docutils literal notranslate"><span class="pre">logs/</span></code> and temp for the path specified for <code class="docutils literal notranslate"><span class="pre">tempDir</span></code> in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>.</p>
<p>We have provided two convenience scripts in 54gene-wgs-germline repository to execute the workflow in a cluster environment, <code class="docutils literal notranslate"><span class="pre">run.sh</span></code> and <code class="docutils literal notranslate"><span class="pre">wrapper.sh</span></code>. The <code class="docutils literal notranslate"><span class="pre">wrapper.sh</span></code> script embeds the <code class="docutils literal notranslate"><span class="pre">snakemake</span></code> command and other command-line flags to control submission of jobs to an HPC using the <code class="docutils literal notranslate"><span class="pre">cluster_mode</span></code> string pulled from the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>. This script also directs all stdout from Snakemake to a log file named <code class="docutils literal notranslate"><span class="pre">WGS_${DATE}.out</span></code> which will include the latest git tag and version of the pipeline, if cloned from our repository. For additional logging information, see <a class="reference internal" href="#logging"><span class="std std-ref">Logging</span></a>.</p>
<p>This wrapper script can be edited to your needs and run using <code class="docutils literal notranslate"><span class="pre">bash</span> <span class="pre">run.sh</span></code>.</p>
</section>
<section id="logging">
<span id="id2"></span><h2>Logging<a class="headerlink" href="#logging" title="Permalink to this headline"></a></h2>
<p>All job-specific logs will be directed to a <code class="docutils literal notranslate"><span class="pre">logs/</span></code> subdirectory in the home analysis directory of the pipeline. This directory automatically created for you by the pipeline.</p>
<p>If you choose to use the <code class="docutils literal notranslate"><span class="pre">wrapper.sh</span></code> script provided and modified for your environment, a <code class="docutils literal notranslate"><span class="pre">WGS_${DATE}.out</span></code> log file containing all stdout from snakemake will also be available in the home analysis directory.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation and Setup" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="investigateresults.html" class="btn btn-neutral float-right" title="Investigate the Results" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, B. Ballew, C. Palmer, E. Joshi, 54gene.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>